vm/boolean.go

```diff
@@ -44,144 +37,160 @@ func (b *BooleanObject) equal(e *BooleanObject) bool {
+			},
 		},
-	},
-	{
-		// Reverse the receiver.
-		//
-		// ```ruby
-		// !true  # => false
-		// !false # => true
-		// ```
-		// @return [Boolean]
-		Name: "!",
-		Fn: func(receiver Object) builtinMethodBody {
-			return func(t *thread, args []Object, blockFrame *callFrame) Object {
-
-				rightValue := receiver.(*BooleanObject).Value
-
-				if rightValue {
+		{
+			// Returns true if the receiver is not equals to the argument.
+			//
+			// ```Ruby
+			// 4 != 2 # => true
+			// 45 != 45 # => false
+			// ```
+			// @return [Boolean]
+			Name: "!=",
+			Fn: func(receiver Object) builtinMethodBody {
+				return func(t *thread, args []Object, blockFrame *callFrame) Object {
+
+					err := checkArgumentLen(args, receiver.returnClass(), "!=")
+					if err != nil {
+						return err
+					}
+
+					if receiver != args[0] {
+						return TRUE
+					}
 					return FALSE
 				}
-
-				return TRUE
-			}
+			},
 		},
-	},
-	{
-		// Returns true if both the receiver and the argument are true.
-		//
-		// ```ruby
-		// 3 > 2 && 5 > 3  # => true
-		// 3 > 2 && 5 > 10 # => false
-		// ```
-		// @return [Boolean]
-		Name: "&&",
-		Fn: func(receiver Object) builtinMethodBody {
-			return func(t *thread, args []Object, blockFrame *callFrame) Object {
-
-				leftValue := receiver.(*BooleanObject).Value
-				right, ok := args[0].(*BooleanObject)
-
-				if !ok {
-					return wrongTypeError(booleanClass)
-				}
-
-				rightValue := right.Value
+		{
+			// Reverse the receiver.
+			//
+			// ```ruby
+			// !true  # => false
+			// !false # => true
+			// ```
+			// @return [Boolean]
+			Name: "!",
+			Fn: func(receiver Object) builtinMethodBody {
+				return func(t *thread, args []Object, blockFrame *callFrame) Object {
+
+					rightValue := receiver.(*BooleanObject).Value
+
+					if rightValue {
+						return FALSE
+					}
 
-				if leftValue && rightValue {
 					return TRUE
 				}
-
-				return FALSE
-			}
+			},
 		},
-	},
-	{
-		// Returns true either if the receiver or argument is true.
-		//
-		// ```ruby
-		// 3 > 2 || 5 > 3  # => true
-		// 3 > 2 || 5 > 10 # => true
-		// 2 > 3 || 5 > 10 # => false
-		// ```
-		// @return [Boolean]
-		Name: "||",
-		Fn: func(receiver Object) builtinMethodBody {
-			return func(t *thread, args []Object, blockFrame *callFrame) Object {
-
-				leftValue := receiver.(*BooleanObject).Value
-				right, ok := args[0].(*BooleanObject)
-
-				if !ok {
-					return wrongTypeError(booleanClass)
-				}
-
-				rightValue := right.Value
+		{
+			// Returns true if both the receiver and the argument are true.
+			//
+			// ```ruby
+			// 3 > 2 && 5 > 3  # => true
+			// 3 > 2 && 5 > 10 # => false
+			// ```
+			// @return [Boolean]
+			Name: "&&",
+			Fn: func(receiver Object) builtinMethodBody {
+				return func(t *thread, args []Object, blockFrame *callFrame) Object {
+
+					leftValue := receiver.(*BooleanObject).Value
+					right, ok := args[0].(*BooleanObject)
+
+					if !ok {
+						return wrongTypeError(receiver.returnClass())
```

Comment:
I just found that we can pass receiver's class directly instead of specifying the class. My bad.